### PR TITLE
tfm: Harden and Enforce TF-M Initial Attestion Key

### PR DESCRIPTION
**Pull request by joerchan**
Enforce that the initial attestation partition has the required
initial attestation key provisioned.

If the initial attestation key (IAK) is not present during boot of
TF-M the system will panic during initialization.

Introduce Kconfig option in zephyr build system that reflects the TF-M
cmake config variable with the same default value for dummy provisioning
and have it satisfy the IAK present requirement.
This configuration is not suitable for production, and by having this
in zephyr configuration we can have this as part of the hardened
configuration check.

### CODE REVIEW

**Comment by joerchan** [comment]
@d3zd3z @microbuilder The issues discovered in this PR appears to be existing in main branch as well.

Would you be able to have a look at this?

**Comment by MaureenHelm** [comment]
@d3zd3z please take a look

**Review by microbuilder — DISMISSED** [review]

**Comment by joerchan** [comment]
Rebased the PR. Hopefully the CI issues are gone now.

**Review by microbuilder — DISMISSED** [review]

File: modules/trusted-firmware-m/Kconfig.tfm [hunk 1]
```diff
@@ -177,6 +177,28 @@ config TFM_PARTITION_PLATFORM_CUSTOM_REBOOT
 	  Instead the application will have to override the weak ARM
 	  implementation of sys_arch_reset().
 
+config TFM_DUMMY_PROVISIONING
+	bool "Provision with dummy values. NOT to be used in production"
+	select TFM_INITIAL_ATTESTATION_KEY
+	default n if BOARD_B_U585I_IOT02A
```
Comment (joerchan) [inline modules/trusted-firmware-m/Kconfig.tfm:183]: I think I should move this to a defconfig instead.

File: modules/trusted-firmware-m/Kconfig.tfm [hunk 2]
```diff
@@ -177,6 +177,28 @@ config TFM_PARTITION_PLATFORM_CUSTOM_REBOOT
...
+	  insecure.
+
+config TFM_INITIAL_ATTESTATION_KEY
+	bool
+	help
+	  Hidden option to mark that the TF-M platform has an initial
+	  attestation key, which is a requirement for the Initial Attestation
+	  partition.
+	  A TF-M platform that supports the usage of Initial Attestation Key
+	  shall enable this configuration.
```
Comment (SebastianBoe) [inline modules/trusted-firmware-m/Kconfig.tfm:200]: Is it a symbol that indicates whether the HW/SDK supports such a key. Of a symbol that indicates whether it is present in the current build?

I feel these two sentences are communicating different things.

File: modules/trusted-firmware-m/Kconfig.tfm (hunk 1 above)
Comment (SebastianBoe) [reply modules/trusted-firmware-m/Kconfig.tfm:183]: Did you forget to delete this line?

I see it is in a defconfig.

File: modules/trusted-firmware-m/CMakeLists.txt [hunk 3]
```diff
@@ -95,6 +95,8 @@ if (CONFIG_BUILD_WITH_TFM)
     list(APPEND TFM_CMAKE_ARGS -DMCUBOOT_IMAGE_NUMBER=${CONFIG_TFM_MCUBOOT_IMAGE_NUMBER})
   endif()
 
+  list(APPEND TFM_CMAKE_ARGS -DTFM_DUMMY_PROVISIONING=${CONFIG_TFM_DUMMY_PROVISIONING})
```
Comment (SebastianBoe) [inline modules/trusted-firmware-m/CMakeLists.txt:98]: Why is this different from the line below ?

Won't CONFIG_TFM_DUMMY_PROVISIONING be undefined when disabled?

so we get

-DTFM_DUMMY_PROVISIONING=

**Review by SebastianBoe — CHANGES_REQUESTED** [review]

File: modules/trusted-firmware-m/Kconfig.tfm (hunk 1 above)
Comment (joerchan) [reply modules/trusted-firmware-m/Kconfig.tfm:183]: No, this is a self-review comment / reminder.

The defconfig is a different symbol, but the defaults are related.

File: modules/trusted-firmware-m/Kconfig.tfm (hunk 2 above)
Comment (joerchan) [reply modules/trusted-firmware-m/Kconfig.tfm:200]: Whether the HW / SDK supports this key. Given that self-provisioning of the IAK is not something that should be done according PSA, this means that it is not really part of the build, but part of the whole "system" of this platforms build.
Like a provisioning image for example.

**Comment by microbuilder** [comment]
> Whether the HW / SDK supports this key. Given that self-provisioning of the IAK is not something that should be done according PSA, this means that it is not really part of the build, but part of the whole "system" of this platforms build.

@joerchan Can you point me to where it's mentioned that self-provisioning is something to avoid? I've never looked in detail, so just started poking around to understand the expected workflow.

In this PSA document on attestation, I see:

> Source: https://arm-software.github.io/psa-api/attestation/1.0/IHI0085-PSA_Certified_Attestation_API-1.0.3.pdf

```
During such an enrolment process, a device might generate a new key pair
and create a Certificate Signing Request (CSR) or equivalent, containing:
∙ The public key of the generated key-pair.
∙ A proof of possession of the corresponding private key (in general this is
   the public key signed by the private key). This protects against
   man-in-the-middle attacks where an attacker can hijack the enrolment
   to insert their own public key into the device request.
∙ An initial attestation, in order for the recipient to assess how that
   particular combination of hardware and firmware can be trusted.

The CSR is then passed to a Certification Authority who can assign it an identity with the new service and
then return an identity certificate signed using the private key of the Certification Authority. The
Certification Authority may be operated by the company who owns the devices or operated by a trusted
third party. Creating extra identities on devices is expected to be a routine operation.
```

A 3rd party key generated off device and added during provisioning also seems normal -- since you need to get the public key into a database as well to verify token signatures -- but I didn't see a statement that self-provisioning (such as deriving a new key from the HUK) should be avoided?

File: modules/trusted-firmware-m/Kconfig.tfm (hunk 1 above)
Comment (joerchan) [reply modules/trusted-firmware-m/Kconfig.tfm]: Updated.

File: modules/trusted-firmware-m/Kconfig.tfm (hunk 2 above)
Comment (joerchan) [reply modules/trusted-firmware-m/Kconfig.tfm]: Removed second sentence.

**Comment by joerchan** [comment]
> joerchan Can you point me to where it's mentioned that self-provisioning is something to avoid? I've never looked in detail, so just started poking around to understand the expected workflow.

@microbuilder It's not that they something to avoid, but provisioning by a different image is something they recommend, and it is our current documented approach.
https://tf-m-user-guide.trustedfirmware.org/integration_guide/platform_provisioning.html#provisioning-production-hardware

**Review by SebastianBoe — APPROVED** [review]

**Review by microbuilder — APPROVED** [review]

**Comment by erwango** [comment]
Jumping out of the blue and possibly a dumb question, but I had a try and (I can confirm this doesn't break anything but) I'm still seeing the following warning when running TF-M samples:
```
[WRN] This device was provisioned with dummy keys. This device is NOT SECURE
```
Is that expected ?

**Review by d3zd3z — APPROVED** [review]

**Comment by joerchan** [comment]
> Jumping out of the blue and possibly a dumb question, but I had a try and (I can confirm this doesn't break anything but) I'm still seeing the following warning when running TF-M samples:
> 
> ```
> [WRN] This device was provisioned with dummy keys. This device is NOT SECURE
> ```
> 
> Is that expected ?

Yes, pretty much all of the boards are dummy provisioned, for the nordic boards you need to use our SDK in addition to get it provisioned.

**Pull request merged**